rclpy/src/rclpy/_rclpy_logging.cpp

```diff
@@ -0,0 +1,193 @@
+
+namespace py = pybind11;
+
+#include <rcutils/allocator.h>
+#include <rcutils/error_handling.h>
+#include <rcutils/logging.h>
+#include <rcutils/time.h>
+
+#include <rcl_logging_interface/rcl_logging_interface.h>
+
+#include <string>
+
+/// Initialize the logging system.
+/**
+ * \return None
+ */
+void
+rclpy_logging_initialize()
+{
+  rcutils_ret_t ret = rcutils_logging_initialize();
+  if (ret != RCUTILS_RET_OK) {
+    rcutils_reset_error();
+    throw std::runtime_error("Failed to initialize logging system");
+  }
+}
+
+/// Shutdown the logging system.
+/**
+ * \return None
+ */
+void
+rclpy_logging_shutdown()
+{
+  // TODO(dhood): error checking
+  rcutils_ret_t ret = rcutils_logging_shutdown();
+  if (ret != RCUTILS_RET_OK) {
+    rcutils_reset_error();
+    throw std::runtime_error("Failed to shutdown logging system");
+  }
+}
+
+/// Set the level of a logger.
+/**
+ *
+ * \param[in] name Fully-qualified name of logger.
+ * \param[in] level to set
+ * \return None
+ */
+void
+rclpy_logging_set_logger_level(const char * name, int level)
+{
+  rcutils_ret_t ret = rcutils_logging_set_logger_level(name, level);
+  if (ret != RCUTILS_RET_OK) {
+    rcutils_reset_error();
+    throw std::runtime_error("Failed to set level for logger");
+  }
+}
+
+/// Get the effective level of a logger.
+/**
+ * The "effective" logger level is determined as the logger level if it has been set explicitly,
+ * otherwise it defers to the level of the logger's ancestors, and if all are unset
+ * the default level is used.
+ *
+ * \param[in] name Fully-qualified name of logger.
+ * \return The effective level
+ */
+int
+rclpy_logging_get_logger_effective_level(const char * name)
+{
+  int logger_level = rcutils_logging_get_logger_effective_level(name);
+
+  if (logger_level < 0) {
+    rcutils_reset_error();
+    throw std::runtime_error("Failed to get effective level for logger");
+  }
+  return logger_level;
+}
+
+/// Determine if the logger is enabled for a severity.
+/**
+ *
+ * \param[in] name Fully-qualified name of logger.
+ * \param[in] severity Logging severity to compare against.
+ * \return True if the logger is enabled for the severity,
+ * \return False otherwise.
+ */
+bool
+rclpy_logging_logger_is_enabled_for(const char * name, int severity)
+{
+  return rcutils_logging_logger_is_enabled_for(name, severity);
+}
+
+/// Log a message through rcutils with the specified severity.
+/**
+ *
+ * \param[in] severity Enum of type RCUTILS_LOG_SEVERITY.
+ * \param[in] name Name of logger.
+ * \param[in] message String to log.
+ * \param[in] function_name String with the function name of the caller.
+ * \param[in] file_name String with the file name of the caller.
+ * \param[in] line_number Line number of the calling function.
+ * \return None
+ */
+void
+rclpy_logging_rcutils_log(
+  int severity,
+  const char * name,
+  const char * message,
+  const char * function_name,
+  const char * file_name,
+  uint64_t line_number)
+{
+  RCUTILS_LOGGING_AUTOINIT;
```

Comment:
Does this require special consideration since it initializes a global variable? Really, the entire logging subsystem for rcutils seems to rely on globals. Does that jive with the Pybind11 model?

Reply:
That's a good question. The only time I know of that global state matters is with subinterpreters. If a program used two python sub-interpreters and imported this module in both, then logging being initialized in one sub-interpreter would mean it's initialized in the other too. The same issue would come up with changes to the other globals like [the output format](https://github.com/ros2/rcutils/blob/d413e135d9d2a520f84b70ce3b0fd9625592ec50/src/logging.c#L88), and [the output handler](https://github.com/ros2/rcutils/blob/d413e135d9d2a520f84b70ce3b0fd9625592ec50/src/logging.c#L94).

It doesn't seem to be a problem because Pybind11 [hardcodes the module size to -1](https://github.com/pybind/pybind11/blob/c0fbb02c9fc1fe969f2740253469499385e6dc20/include/pybind11/pybind11.h#L1041), meaning Pybind11 modules [can't support subinterpreters](https://docs.python.org/3/c-api/module.html#c.PyModuleDef.m_size) anyways.

Reply:
Thanks for explaining. Seems like a non-issue, then.